### PR TITLE
[`Cont'd`] Add the SDE variant of ~~DPM-Solver~~ and DPM-Solver++ to DPM Single Step

### PR DESCRIPTION
**Pull request by tolgacangoz**
Thanks for the opportunity to continue #4251 @yiyixuxu.

There was uncertainty about whether `sigma_s0` or `sigma_s1` should be used in the second-order update in `sde-dpmsolver++` when @cmdr2 was implementing. I tried both and chose `sigma_s1`:
`sigma_s0`             |  `sigma_s1`
:-------------------------:|:-------------------------:
| sigma_s0 | sigma_s1 |

[Reproduction notebook.](https://gist.github.com/tolgacangoz/2c7f8eec02ea532f6cd742c004a47909)

@LuChengTHU I want to congratulate you on this invention and thank you for your previous DPM-Solver contributions in the repo. It would be very nice to review this version. This PR is the exact copy of your "Add the SDE variant of DPM-Solver and DPM-Solver++ [for multi-step]" #3344 PR except `sigma_s1` is used in the second-order update rather than `sigma_s0`.

### CODE REVIEW

**Comment by LuChengTHU** [comment]
Thanks for this nice PR! For single-step method we indeed should use `sigma_s1` because we always take a "large step" for single-step method and re-use some intermediate variables (i.e., at time `s0`). So I believe this PR is correct!

**Comment by tolgacangoz** [comment]
Thank you very much for your review!

**Comment by tolgacangoz** [comment]
`sde-dpmsolver++` seems reasonable; but `sde-dpmsolver` creates almost complete noise:
```py
!git clone -b sde-dpmsolver-single-step https://github.com/tolgacangoz/diffusers.git
%cd diffusers
!pip install -e . accelerate -q

from diffusers import AutoPipelineForText2Image, DPMSolverSinglestepScheduler
import torch

pipeline_text2image = AutoPipelineForText2Image.from_pretrained(
    "stabilityai/stable-diffusion-xl-base-1.0", torch_dtype=torch.float16, variant="fp16",
    use_safetensors=True).to("cuda")
prompt = "Astronaut in a jungle, cold color palette, muted colors, detailed, 8k"
generator = torch.Generator('cpu').manual_seed(0)
pipeline_text2image.scheduler = DPMSolverSinglestepScheduler.from_config(
    pipeline_text2image.scheduler.config, algorithm_type='sde-dpmsolver',
    final_sigmas_type='sigma_min')
image = pipeline_text2image(prompt=prompt, generator=generator).images[0]
image
```
image

`sde-dpmsolver` for `DPMSolverMultistepScheduler` also creates considerably noisy outputs; but not like complete noise.

**Comment by tolgacangoz** [comment]
For `sde-dpmsolver`:

Multi-step `num_inference_steps=50`            |  Single-step `num_inference_steps=50`
:-------------------------:|:-------------------------:
| m_50 | s_50 |

Multi-step `num_inference_steps=100`            |  Single-step `num_inference_steps=100`
:-------------------------:|:-------------------------:
| m_100 | s_100 |

Even multi-step with `num_inference_steps=100` seems noisy. Tbh, my math understanding level is insufficient to understand this study well.
@LuChengTHU How should we conclude about `sde-dpmsolver` implementation?

**Comment by LuChengTHU** [comment]
Hi @tolgacangoz , I would suggest only use sde-dpmsolver++ because this formulation is more suitable for SDEs and can include stochastic DDIM / DDPM as the first order case. For sde-dpmsolver it may even need 1000 steps to converge so it is hard to use in practice.

**Comment by tolgacangoz** [comment]
> Hi @tolgacangoz , I would suggest only use sde-dpmsolver++ because this formulation is more suitable for SDEs and can include stochastic DDIM / DDPM as the first order case. For sde-dpmsolver it may even need 1000 steps to converge so it is hard to use in practice.

OK, thank you for your feedback!

**Comment by tolgacangoz** [comment]
@yiyixuxu

**Comment by tolgacangoz** [comment]
@yiyixuxu, will we continue [with this](https://github.com/huggingface/diffusers/issues/5092#:~:text=DPMSolverSDEScheduler,deprecate%20this%20pipeline)?

**Comment by tolgacangoz** [comment]
@yiyixuxu, this is ready for your review now.
Will one continue [with this](https://github.com/huggingface/diffusers/issues/5092#:~:text=DPMSolverSDEScheduler,deprecate%20this%20pipeline)?

**Comment by yiyixuxu** [comment]
@LuChengTHU is this PR good to merge now?

File: src/diffusers/schedulers/scheduling_dpmsolver_singlestep.py [hunk 1]
```diff
@@ -517,34 +518,43 @@ def convert_model_output(
                 x0_pred = self._threshold_sample(x0_pred)
 
             return x0_pred
+
         # DPM-Solver needs to solve an integral of the noise prediction model.
-        elif self.config.algorithm_type == "dpmsolver":
+        elif self.config.algorithm_type in ["dpmsolver", "sde-dpmsolver"]:
```
Comment (yiyixuxu) [inline src/diffusers/schedulers/scheduling_dpmsolver_singlestep.py:523]: should we not support `"sde_dpmsolver" here if it does not work?

Comment (tolgacangoz) [reply src/diffusers/schedulers/scheduling_dpmsolver_singlestep.py:523]: @LuChengTHU says `sde_dpmsolver` is hard to use in practice and you are unwilling to support it; thus I will remove it.

Comment (yiyixuxu) [reply src/diffusers/schedulers/scheduling_dpmsolver_singlestep.py:523]: ok let's remove it then
we can merge this after that:)

**Comment by yiyixuxu** [comment]
@tolgacangoz 
thank you very much! this is a very nice PR :)

**Comment by tolgacangoz** [comment]
Thanks for merging and labeling as a good PR!

**Pull request merged**